Firestore/core/src/firebase/firestore/auth/firebase_credentials_provider_apple.mm

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/auth/firebase_credentials_provider_apple.h"
+
+#import <FirebaseCore/FIRApp.h>
+#import <FirebaseCore/FIRAppInternal.h>
+#import <FirebaseCore/FIROptionsInternal.h>
+
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+
+namespace firebase {
+namespace firestore {
+namespace auth {
+
+std::mutex FirebaseCredentialsProvider::mutex_;
+
+FirebaseCredentialsProvider::FirebaseCredentialsProvider()
+    : FirebaseCredentialsProvider([FIRApp defaultApp]) {
+}
+
+FirebaseCredentialsProvider::FirebaseCredentialsProvider(FIRApp* app)
+    : app_(app),
+      current_user_(firebase::firestore::util::MakeStringView([app getUID])),
+      user_counter_(0) {
+  auth_listener_handle_ = [[NSNotificationCenter defaultCenter]
+      addObserverForName:FIRAuthStateDidChangeInternalNotification
+                  object:nil
+                   queue:nil
+              usingBlock:^(NSNotification* notification) {
+                std::unique_lock<std::mutex> lock(mutex_);
+                NSDictionary* user_info = notification.userInfo;
```

Comment:
Please use generics instead of raw types. I think this is `NSDictionary<NSString *, id> *`